src/field.rs

```diff
@@ -200,78 +200,86 @@ pub trait FieldElementWithInteger: FieldElement + From<Self::Integer> {
         }
-
-        // If `i` is still not zero, this means that it cannot be encoded by `bits` bits.
-        if i != Self::zero_integer() {
+        if input >> bits != Self::zero_integer() {
             return Err(FieldError::InputSizeMismatch);
         }
 
-        Ok(())
+        Ok(BitvectorRepresentationIter {
+            inner: 0..bits,
+            input,
+        })
     }
 
-    /// Encode `input` as `bits`-bit vector of elements of `Self` if it's small enough
-    /// to be represented with that many bits.
+    /// Inverts the encoding done by [`Self::encode_into_bitvector_representation`], and returns a
+    /// single field element.
     ///
-    /// # Arguments
+    /// This performs an inner product between the input vector of field elements and successive
+    /// powers of two (starting with 2^0 = 1). If the input came from
+    /// [`Self::encode_into_bitvector_representation`], then the result will be equal to the
+    /// originally encoded integer, projected into the field.
     ///
-    /// * `input` - The field element to encode
-    /// * `bits` - The number of bits to use for the encoding
-    fn encode_into_bitvector_representation(
-        input: &Self::Integer,
-        bits: usize,
-    ) -> Result<Vec<Self>, FieldError> {
-        let mut result = vec![Self::zero(); bits];
-        Self::fill_with_bitvector_representation(input, &mut result)?;
-        Ok(result)
-    }
-
-    /// Decode the bitvector-represented value `input` into a simple representation as a single
-    /// field element.
+    /// Note that this decoding operation is linear, so it can be applied to secret shares of an
+    /// encoded integer, and if the results are summed up, it will be equal to the encoded integer.
     ///
     /// # Errors
     ///
     /// This function errors if `2^input.len() - 1` does not fit into the field `Self`.
     fn decode_from_bitvector_representation(input: &[Self]) -> Result<Self, FieldError> {
-        let fi_one = Self::one_integer();
-
         if !Self::valid_integer_bitlength(input.len()) {
             return Err(FieldError::ModulusOverflow);
         }
 
         let mut decoded = Self::zero();
-        for (l, bit) in input.iter().enumerate() {
-            let w = fi_one << l;
-            decoded += Self::from(w) * *bit;
+        let one = Self::one();
+        let two = one + one;
+        let mut power_of_two = one;
+        for value in input.iter() {
+            decoded += *value * power_of_two;
+            power_of_two *= two;
         }
         Ok(decoded)
     }
+}
+
+/// This iterator returns a sequence of field elements that are equal to zero or one, representing
+/// some integer in two's complement form. See
+/// [`FieldElementWithInteger::encode_into_bitvector_representation`].
+// Note that this is implemented with a separate struct, instead of using the map combinator,
+// because return_position_impl_trait_in_trait is not yet stable.
+#[derive(Debug)]
```

Comment:
Any reason not to derive `Clone` here? I feel like I always regret not deriving `Clone` if it's trivial to do so.